source/extensions/tracers/datadog/span.cc

```diff
@@ -0,0 +1,118 @@
+#include "source/extensions/tracers/datadog/span.h"
+
+#include <datadog/dict_writer.h>
+#include <datadog/sampling_priority.h>
+#include <datadog/span_config.h>
+#include <datadog/trace_segment.h>
+
+#include <utility>
+
+#include "source/common/tracing/null_span_impl.h"
+#include "source/extensions/tracers/datadog/time_util.h"
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Datadog {
+namespace {
+
+class TraceContextWriter : public datadog::tracing::DictWriter {
+public:
+  explicit TraceContextWriter(Tracing::TraceContext& context) : context_(&context) {}
+
+  void set(datadog::tracing::StringView key, datadog::tracing::StringView value) override {
+    context_->setByKey(key, value);
+  }
+
+private:
+  Tracing::TraceContext* context_;
+};
+
+} // namespace
+
+Span::Span(datadog::tracing::Span&& span) : span_(std::move(span)) {}
+
+const datadog::tracing::Optional<datadog::tracing::Span>& Span::impl() const { return span_; }
+
+void Span::setOperation(absl::string_view operation) {
+  if (!span_) {
+    return;
+  }
+
+  span_->set_name(operation);
+}
+
+void Span::setTag(absl::string_view name, absl::string_view value) {
+  if (!span_) {
+    return;
+  }
+
+  span_->set_tag(name, value);
+}
+
+void Span::log(SystemTime, const std::string&) {
+  // Datadog spans don't have in-bound "events" or "logs".
+}
+
+void Span::finishSpan() { span_.reset(); }
+
+void Span::injectContext(Tracing::TraceContext& trace_context,
+                         const Upstream::HostDescriptionConstSharedPtr&) {
+  if (!span_) {
+    return;
+  }
+
+  TraceContextWriter writer{trace_context};
+  span_->inject(writer);
+}
+
+Tracing::SpanPtr Span::spawnChild(const Tracing::Config&, const std::string& name,
+                                  SystemTime start_time) {
+  if (!span_) {
+    // I don't expect this to happen. This means that `spawnChild` was called
+    // after `finishSpan`.
```

Comment:
should we use an ENVOY_BUG here (logs a message in production build, asserts in debug/default builds)?

Reply:
That might be worth doing.

I'd reserve `assert` and friends for cases where the detected behavior is in violation of the component's contract. Because this `Span` has the escape hatch of going into "null-mode," it's not necessary. By that reasoning, perhaps the code comment is not necessary either.

You mentioned the idea of @moderation looking over the behavior of this `Span`. Maybe they can decide whether "spawn-after-finish" ought to be considered forbidden in these tracing implementations.